File: 2684-maximum-number-of-moves-in-a-grid/2684-maximum-number-of-moves-in-a-grid.java
class Solution {
    public int[] dy = {-1, 0, 1};
    public int[][] dp;
    public int h;
    public int w;
    public int ans;
    
    public boolean outOfRange(int y, int x) {
        return y < 0 || y >= h || x < 0 || x >= w;
    }
    
    public int dfs(int y, int x, int[][] grid) {         
        if (dp[y][x] != -1) {
            return dp[y][x];
        }
        
        dp[y][x] = 0;
        
        for (int i = 0; i < 3; i++) {
            int ny = y + dy[i];
            int nx = x + 1;
            
            if (outOfRange(ny, nx) || grid[ny][nx] <= grid[y][x]) {
                continue;
            }
            
            dp[y][x] = Math.max(dp[y][x], dfs(ny, nx, grid));
        }
        
        return dp[y][x] + 1;
    }
    
    public int maxMoves(int[][] grid) {
        h = grid.length;
        w = grid[0].length;
        dp = new int[h][w];
        
        for (int i = 0; i < h; i++) {
            Arrays.fill(dp[i], -1);
        }
        
        int ans = 0;
        for (int i = 0; i < h; i++) {
            ans = Math.max(ans, dfs(i, 0, grid));
        }
        
        return ans - 1;
    }
}

File: 2684-maximum-number-of-moves-in-a-grid/2684-maximum-number-of-moves-in-a-grid.cpp
class Solution {
public:
    typedef vector<int>vi;
    typedef vector<vi>vii;
    int maxMoves(vector<vector<int>>& grid) {
        int h=grid.size();
        int w=grid.size();
        vii dp=vii(h+2,vi(w+1,0));
        vii g=vii(h+2,vi(w+1,0));
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[i].size();j++){
                g[i+1][j+1]=grid[i][j];
            }
        }
        for(int i=1;i<=h;i++){
            dp[i][1]=1;
        }
        int answer=1;
        for(int i=2;i<=w;i++){
            for(int j=1;j<=h;j++){
                int ma=0;
                for(int move=-1;move<=1;move++){
                    if(g[j+move][i-1]>=g[j][i]){
                        continue;
                    }
                    ma=max(ma,dp[j+move][i-1]);
                }
                if(ma==0){
                    continue;
                }
                dp[j][i]=max(dp[j][i],ma+1);
                answer=max(answer,dp[j][i]);
            }
        }
        return answer-1;
    }
};